In [101]:
import cv2
import numpy as np

In [2]:
img = cv2.imread(r'C:\Users\Lenovo\Desktop\open_cv_data\Exercise Files\Ch02\02_01 Begin\opencv-logo.png')

img.shape

(739, 600, 3)

We see that the image has 739 pixels with 600 columns, and has 3 (bgr) channels.

In [3]:
img.dtype

dtype('uint8')

Note that the data type of img is 'uint8' which indicates the maximum values of a $2^8$ values in each pixel, i.e. the range of a pixel is $2^8$

In [6]:
# we can open a window to display the image with namedWindow
cv2.namedWindow('Image',cv2.WINDOW_NORMAL) 
cv2.imshow('Image', img)

# add a wait element so that the image will hang 
cv2.waitKey(0)

# we can call destroyAllWindows which closes the window when we press on any key 
cv2.destroyAllWindows()

In [7]:
# we can store image with imwrite. we can even change the file type such as from png to jpg
cv2.imwrite('image2.jpg', img)

True

Let's create a black cell with `np.zeros` and show it in cv2:

In [8]:
black = np.zeros([140, 200, 3], 'uint8')

cv2.namedWindow('Black', cv2.WINDOW_NORMAL)

cv2.waitKey(0)
cv2.destroyAllWindows()

In [9]:
ones = np.ones([150, 110, 3], 'uint8')
white = ones*(2**8-1)

cv2.namedWindow('White', cv2.WINDOW_NORMAL)
cv2.imshow('White', white)

cv2.waitKey(0)
cv2.destroyAllWindows()

Let's now create a blue image:

In [10]:
blue = ones.copy()
blue[:, :] = (255, 0, 0)

cv2.namedWindow('Blue', cv2.WINDOW_NORMAL)
cv2.imshow('Blue', blue)

cv2.waitKey(0)
cv2.destroyAllWindows()

BGR TO HSV transformation:

In [11]:
butterfly = cv2.imread(r'C:\Users\Lenovo\Desktop\open_cv_data\Exercise Files\Ch02\02_05 Begin\butterfly.jpg', 1)

cv2.imshow('BF', butterfly)
cv2.moveWindow('BF', 0, 0)

b, g, r = cv2.split(butterfly)
height, width, channel = butterfly.shape

rgb_split = np.empty([height, width*3, channel], 'uint8')

rgb_split[:, 0:width] = cv2.merge([b, b, b])
rgb_split[:, width:2*width] = cv2.merge([g, g, g])
rgb_split[:, 2*width:3*width] = cv2.merge([r, r, r])

cv2.imshow('Split', rgb_split)
cv2.moveWindow('Split', 0, height)

# HSV Split
hsv_split = cv2.cvtColor(butterfly, cv2.COLOR_BGR2HSV)
h, s, v = cv2.split(hsv_split)
hsv_split = np.concatenate([h, s, v], axis=1)
cv2.imshow('HSV', hsv_split)

cv2.waitKey(0)
cv2.destroyAllWindows()

In [12]:
hsv_split.shape

(356, 1479)

Color to Gray:

In [13]:
gray = cv2.cvtColor(butterfly, cv2.COLOR_RGB2GRAY)
cv2.imwrite('gray_butterfly.jpg', gray)

True

Adding transparency to the figure:

In [14]:
b, g, r = [butterfly[:, :, i] for i in range(3)]

rgba = cv2.merge((b, g, r, g))
cv2.namedWindow('RGBA', cv2.WINDOW_NORMAL)
cv2.imshow('rgba', rgba)

cv2.imwrite('rgba_butterfly.png', rgba)
cv2.waitKey(0)
cv2.destroyAllWindows()

**Blur, Dilation, Erotion**

In [23]:
thresh = cv2.imread(r'C:\Users\Lenovo\Desktop\open_cv_data\Exercise Files\Ch02\02_06 Begin\thresh.jpg')

cv2.imshow('Thresh', thresh)
cv2.moveWindow('Thresh', 0, 0)

height, width = thresh.shape[0], thresh.shape[1]

#Let's gaussian blur the image respectively on y and x axis. Note that gaussian blur coefficients need to be odd numbers.

blur_y = cv2.GaussianBlur(thresh, (1, 33), 0)
cv2.imshow('Blur_y', blur_y)
cv2.moveWindow('Blur_y', thresh.shape[1], 1)

blur_x = cv2.GaussianBlur(thresh, (33, 1), 0)
cv2.imshow('Blur_x', blur_x)
cv2.moveWindow('Blur_x', 2*thresh.shape[1], 1)

cv2.waitKey(0)
cv2.destroyAllWindows()

In [30]:
# Now, we will dilate and erode images. Axis lengths still need to be odd numbers 

kernel = np.ones((2, 2), 'uint8')/2

dilate = cv2.dilate(thresh, kernel, iterations=1)
erode = cv2.erode(thresh, kernel, iterations=1)

cv2.imshow('Original', thresh)
cv2.moveWindow('Original', 0, 0)
cv2.imshow('Dilate', dilate)
cv2.moveWindow('Dilate', width, 0)
cv2.imshow('Erode', erode)
cv2.moveWindow('Erode', width*2, 0)

cv2.waitKey(0)
cv2.destroyAllWindows()

Scale the image:

In [75]:
players = cv2.imread(r'C:\Users\Lenovo\Desktop\open_cv_data\Exercise Files\Ch02\02_07 Begin\players.jpg')
cv2.imshow('Players', players)

resize = cv2.resize(players, (0,0), fx=0.5, fy=0.5)
cv2.imshow('Resize', resize)

stretch = cv2.resize(players, (600, 600))
cv2.imshow('Stretch', stretch)

interpolation = cv2.resize(players, (600, 600), cv2.INTER_NEAREST)
cv2.imshow('Interpolation', interpolation)

cv2.waitKey(0)
cv2.destroyAllWindows()

Rotate the image:

In [95]:
# we need to specify the origin of rotation in the matrix additionaly with rotation degree and scale
M = cv2.getRotationMatrix2D((players.shape[1]/2, players.shape[0]/2), 90, 0.5)
rotate = cv2.warpAffine(players, M, (players.shape[1], players.shape[0]))

cv2.imshow('30', rotate)

cv2.waitKey(0)
cv2.destroyAllWindows()

**Video Capture**

In [102]:
cap = cv2.VideoCapture(0)

point = (0, 0)
radius = 100
color = (0, 255, 0)
line_width = 3

while True:
    ret, frame = cap.read()
    cv2.circle(frame, point, radius, color, line_width)
    
    frame = cv2.resize(frame, (0,0), fx=0.5, fy=0.5)
    cv2.imshow('Frame', frame)
    
    ch = cv2.waitKey(1)
    
    if ch & 0XFF == ord('q'):
        break
        
cap.release()
cv2.destroyAllWindows()        

Rotating pictures by manipulating through indexes:

In [103]:
players = cv2.imread(r'C:\Users\Lenovo\Desktop\open_cv_data\Exercise Files\Ch02\02_07 Begin\players.jpg')
cv2.imshow('Players', players)

p = players.copy()
p[:, :] = p[::-1, ::-1]
cv2.imshow('P', p)

p2 = players.copy()
p2[:, :] = p2[:, ::-1]
cv2.imshow('P2', p2)

cv2.waitKey(0)
cv2.destroyAllWindows()

In [115]:
cap = cv2.VideoCapture(0)

point = (0, 0)
radius = 100
color = (0, 0, 255)
line_width = 3

def click(event, x, y, flags, param):
    global point, pressed 
    if event == cv2.EVENT_LBUTTONDOWN:
        print('pressed')
        point = (x, y)
    
cv2.namedWindow('Frame')    
cv2.setMouseCallback('Frame', click)
 
while True:
    ret, frame = cap.read()
    cv2.circle(frame, point, radius, color, line_width)
    
    frame = cv2.resize(frame, (0, 0), fx=1, fy=1)
    cv2.imshow('Frame', frame)
    ch = cv2.waitKey(1)
    if ch & 0XFF == ord('q'):
        break
        
cap.release()        
cv2.destroyAllWindows()

pressed
pressed
pressed


A simple drawing notebook:

In [150]:
cap = cv2.VideoCapture(0)

point = (0, 0)
radius = 5
color = (0,255, 0) 

canvas = np.ones([500, 500, 3], 'uint8')

def click(event, x, y, flags, param):
    global canvas, pressed
    if event == cv2.EVENT_LBUTTONDOWN:
        cv2.circle(canvas, (x,y), radius, color,-1)
        pressed = True
    elif event == cv2.EVENT_MOUSEMOVE and pressed == True:
        cv2.circle(canvas, (x, y), radius, color, -1)
    elif event == cv2.EVENT_LBUTTONUP:
        pressed = False
        c = (0, 0, 255)
        cv2.circle(canvas, (x, y), radius, c, -1) 
    elif event == cv2.EVENT_MOUSEMOVE and pressed == False:
        c = (0, 0, 255)
        cv2.circle(canvas, (x, y), radius, c, -1)

cv2.namedWindow('Canvas')
cv2.setMouseCallback('Canvas', click)
    
while True:
    
    cv2.imshow('Canvas', canvas)
    ch = cv2.waitKey(1)
    if ch & 0XFF == ord('q'):
        break
    
cv2.waitKey(0)     
cv2.destroyAllWindows()        